### DISCUSSION TITLE
task3：使用hardhat初始化生成项目，但是目前提交任务区规范只提交合约文件，是否可以调整优化提交任务规范

### DISCUSSION THREAD

**Discussion by xiangnuans** [openbuildxyz/Web3-Frontend-Bootcamp#1011] [General]
![image](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/assets/24558814/8899c41f-514d-4881-a330-b096de321f18)
task3本来就是使用hardhat初始化的一个项目
会有一些项目初始化配置文件，以及当时使用本地测试链，编写的node执行脚本

但是目前review这按照社区提交任务规范要求删除文件，只保留合约文件，是否合理，欢迎大家积极讨论：

**Comment by ourai** [comment]
在 task 2 的提问「你对去中心化的理解」中[我是这么回答的](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/blob/3989cff6536ce5fe576ff3500b09957c6f75a025/members/ourai/task2/readme.md?plain=1#L58-L62)：

> 顾名思义，「去中心化」就是没有中心，也就不会有中央集权的情况，于是要做任何事需要网络中的每个节点间达成共识，最大化民主。
>
> **但这也造成了决策流程长、效率低下等问题，因而需要就「如何达成共识」这个基本核心问题先达成共识，之后的其他共识按照这个来执行。**

应用到 PR 提交与审核上，就是：

1. 提交规范是什么？作业是什么？
2. 提交人按上述要求提交
3. 审核人按上述要求审核

所以「1」是「基本核心问题」，只要对这个达成共识，后面只需按照这个共识去执行就对了；若有疑义，也是去挑战并推翻或修改这「基本核心问题」。

针对目前的这个问题，提交「源码」这点毋庸置疑，不能提交安装的依赖、编译后文件等；至于「源码」的指代范围，有待商榷。

但是有一点实际问题需要注意到——**提交完整项目会增加仓库整体体积，会产生多大影响要看实际参与提交 task 的人数。**

**Comment by linghuccc** [comment]
1. 本身这个 project，就不应该在建在 task3 文件夹之下，所以原本只需要新建 task3 文件夹，从 project 目录拷贝 3 个 solidity 文件过来即可；
2. 如果非要把 project 建在 task3 文件夹下，其他的文件，并不需要删除，而是可以用 .gitignore 来排除在上传文件之外；
3. 提交的内容应该按照任务规范来，在提交之前，就应该想好如何提交，如果对任务规范有问题，应该在提交之前提出，而不是提交不符合任务规范的内容，被 review 之后，再来质疑任务规范；
4. 如果什么文件都 merge，depository 会很臃肿，充满很多不需要的内容；
5. 如果说 deploy script，project setting 还勉强可以强调一下，是 project 的一部分，那很多人提交的 build info，binary file，就是完全无用的垃圾了。

**Reply by linghuccc** [reply]
请仔细阅读 PR 规范，里面说的很清楚，提交 PR，不是由着你觉得，而是要看 PR 规范是怎么写的

**Reply by ImSingee** [reply]
@lispking 注意，这个要提交的是「作业」而不是完整的「你完成作业的过程」，这两个有本质区别，作业的仓库不是给你用来存档一切使用的，同样，「提交什么」和「自己完成的时候是怎么做的」完全没有必然关系，不要混淆概念混为一谈

**Reply by ImSingee** [reply]
@lispking 你是交作业，不是提交工程；就像是上学的时候，老师让你「撕下来练习册的第 18 页交上来」有问题吗？

**Reply by ImSingee** [reply]
@lispking 首先我只是一个学员

你要知道，你交的是「作业」，而不是一个「可以复现的项目」；项目的目的是可以完整的执行，作业的目的只是让你证明你完成了而已

你正面回答我下面的三个问题

1. 上学的时候，老师让你交寒假作业的其中一页，你会说「老师，这一页的题干在上一页」吗？
2. 申请专利的时候，让你提交相关代码，你会说「这个代码需要 XX 配置才能跑得起来我必须打印所有代码」吗？
3. 写论文的时候，你需要提交相关材料，你会说「这个代码需要 XX 脚本才能跑得起来我必须给你完整的所有代码包」吗？

**作业是检验你证明你做了的，仅此而已**

就这件事情，你可以说其他 task 都要求提交完整项目而这个要求提交特定文件不合理，但你无法单独说这个要求不合理

**Reply by ImSingee** [reply]
@lispking 我发现你的逻辑十分混乱

1. 「既然你是菜鸟」-> 推导出这句话的前提是哪里？不是这个训练营的 mentor 就是菜鸟了吗？退一步，我们讨论的问题，和是否是菜鸟有什么实质关系吗？
2. 「【作业规范】有定义这些吗」-> 我们讨论的不就是作业规范吗？
3. 「老师有站出来说不行吗？并没有」-> 如果老师没有站出来说「不允许提交整个项目」，那么还会有这个讨论吗？
4. 「不懂的话，就别再说话了」 -> 你没有权利让别人闭嘴，反而，你没有正面回复我的三个问题，我有理由认为你的答案就是「不会」，而你不会的原因，则很可能是「因为你知道和他们说了也没用」，你被迫闭嘴了，而你在这里说，因为这里没有人让你闭嘴，但你要知道，**给你张嘴的权利不是让你滥用这个权利的，更不是给你随意冒犯的权利的**
5. 当然还有个问题，你自始至终没有搞明白到底什么是「作业」这个名词的定义，当然，你可能你要重新定义这个名词，这是你的权利，但你需要拿出论点来反驳别人，而不是一味的强调自己的观点，更不是以直接人身攻击冒犯别人的方式诉说自己的观点，这是很情绪化很没有价值的事情


最后，虽然我很想有素质，但是既然你选择了冒犯，那么在你正面逐条有逻辑的回答我的问题之前我也不会再回复你了，因为 **不要和傻子理论，因为他们只会把你拉到他们的层次**

**Reply by linghuccc** [reply]
> @ImSingee 如果是这样的话，说句冒犯的话，那这样的话，**水平就太菜了，你们要不别改作业了，换个专业的老师来改**，合约都在一个文件夹里，这么简单的事情都不会？还来扯让千人同步修改的问题？这样做合适吗？每个人将hardhat工程提交上来，对于仓库来说才占用多少空间？这种问题需要在这里扯半天？
> 
> **那请你们还是别改作业了，换个专业的老师来改！ 那请你们还是别改作业了，换个专业的老师来改！ 那请你们还是别改作业了，换个专业的老师来改！**

你大可以报名当老师，web3 本就是英雄不问出处，但你现在既然是学员，还请你对其他学员和老师尊重一点。这个课程本就是免费课程，还提供奖学金，即使课程内容有所疏漏，作业要求可能有地方不够合理，我相信绝大部分的学员参加这个课程，也是得多于失。
如果你觉得哪里值得改进，欢迎提出意见，但是不等于课程一定要马上按你的要求去改变，毕竟，你也不过是数百名学员中的一员，大家的意见都一样重要。有理不在声高，不是说你到处发消息，到处说别人水平低，就代表你是权威，大家就必须跟着你的指挥棒转。

**Reply by linghuccc** [reply]
> 哈哈，大佬说得是，但是规范不合理就要优化，这种是人制定的，不合理就一起优化，而不是在这里扯这些没用的
> 
> 有一个好处是，后续开下期的时候，可少走弯路，如果大佬不懂这些的话，可以付费，手把手教你
> 
> 这就是一个项目过程改进，review的目的同时是帮忙项目成长，而不是在这扯一些本就不合理的事情，纯浪费大家的时间，这种太低效了，大佬平时是这么写代码的么？如果是，那得改改了

如果你觉得不合理，完全可以提出来，虽然在这个课程里面，你是学员，我是 reviewer，但这并不意味着我的身份比你高，或者我的水平就一定比你高。

web3 最重视的是去中心化，人人都有发声的权利。但要记住，我只是一个人，你也只是一个人，这个课程还有数百名学员的声音，也值得被听取。

强迫别人接受自己的观点，一旦观点不同，就用言语肆意攻击其他同学或 reviewer，是不可取的。

**Reply by linghuccc** [reply]
> 什么叫尊重？让千人去跟着一起去修改，这就是尊重别人吗？这只是一个作业，有工程代码到底有什么问题？这是你先提出来的不合理的建议，作业要求要写着不让用hardhat了吗？课程免费和收费并没有半毛钱关系，课程的本质是带大家入门，让大家有一个很好的交流平台，大家一起共同进步，请问哪里到处发消息？这是一种诬陷行为，有意见就不能讨论？你的意见就是权威？算了，不跟固执的人说太多

讨论问题，就好好讨论。不要随便扣帽子。

“千人”只存在于你的臆想之中，事实上，昨天有 66 个 task3 的 PR，我总共 review 了 65 份（除了我自己的），1/3 是按照 PR 要求提交，并没有错误的，被顺利 merge 了；1/3 有其他错误；因为你说的“工程规范”，被我打回的，只占 1/3，而其中，还有一半是连 build，binary file 一起提交的。也就是说，完全因为你所谓的“工程规范”，真正被我 reject 的，只有十几份。这个数值，跟你动不动就说的“千人”之间的差距，我想你还是会算的。

话说回来，你每次都说让所有学员一起去修改，似乎是个大得不得了的工程，你该不会不知道，在你眼中非常困难，非常“不尊重别人"的工作，只需要几分钟就可以搞定吧？

我从来就没说过我的意见是权威，也从来没说不能讨论，我反对的是用言语攻击不同观点的其他学员的行为。

**Comment by xiangnuans** [comment]
🚄 讨论的范围是：
🚀 **hardhat初始化项目的配置文件和执行脚本**
是否应该提交到task3？

**Comment by ImSingee** [comment]
是否提交工程我觉得都 ok，可能只是为了降低审核人员的压力

不过还有一个模糊的问题是：是否应该提交 openzeppelin 的依赖？

**Reply by linghuccc** [reply]
谈不上压力，因为审核人员根本不会去看，审核人员没有必要去看不在审查范围之内的东西；
openzeppelin 的依赖肯定是不要提交的，这一点也不模糊，非常确定。

**Comment by lispking** [comment]


**Reply by linghuccc** [reply]
是你在纠结工程规范，我 review 是完全按照 PR 规范和要求，跟工程规范无关。

PR 规范相当于试卷/试题，现在是你觉得试卷出的不符合你认为应该的规范，不愿意按照试卷要求作答，非要求改试卷。

**Comment by lispking** [comment]


**Reply by linghuccc** [reply]
规范并没说，一定要用哪种工具，但是说的很清楚，提交合约文件。

写 solidity 合约，现在主要有 4 种框架，truffle, hardhat, foundry 和 brownie。 当然了，你可以把其中 2 种定义为 “常用”。

我从来没说过目录规范会浪费时间，至于什么挑战官方，我只能说你想太多了。

你如果想，可以报名审核合约。你想做什么，那是你的事情，但不要想当然的觉得，别人一定要按你的想法去做。

至于用 hardhat 来写合约，是加分项嘛，见仁见智吧。
